compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesBindingCodeGen.kt

```diff
@@ -0,0 +1,190 @@
+ */
+internal object ContributesBindingCodeGen : AnvilApplicabilityChecker {
+
+  override fun isApplicable(context: AnvilContext) = !context.generateFactoriesOnly
+
+  fun generate(
+    originClass: ClassName,
+    scopes: Iterable<ClassName>,
+  ): FileSpec {
+    val fileName = originClass.generateClassName().simpleName
+    val generatedPackage = HINT_BINDING_PACKAGE_PREFIX +
+      originClass.packageName.safePackageString(dotPrefix = true)
+    val propertyName = originClass.canonicalName.replace('.', '_')
+    return FileSpec.createAnvilSpec(generatedPackage, fileName) {
+      addProperty(
+        PropertySpec
+          .builder(
+            name = propertyName + REFERENCE_SUFFIX,
+            type = KClass::class.asClassName().parameterizedBy(originClass)
+          )
+          .initializer("%T::class", originClass)
+          .addModifiers(PUBLIC)
+          .build()
+      )
+
+      scopes.forEachIndexed { index, scope ->
+        addProperty(
+          PropertySpec
+            .builder(
+              name = propertyName + SCOPE_SUFFIX + index,
+              type = KClass::class.asClassName().parameterizedBy(scope)
+            )
+            .initializer("%T::class", scope)
+            .addModifiers(PUBLIC)
+            .build()
+        )
+      }
+    }
+  }
+
+  internal class KspGenerator(
+    override val env: SymbolProcessorEnvironment,
+  ) : AnvilSymbolProcessor() {
+
+    override fun processChecked(resolver: Resolver): List<KSAnnotated> {
+      resolver
+        .getSymbolsWithAnnotation(contributesBindingFqName.asString())
+        .mapNotNull { annotated ->
+          when {
+            annotated !is KSClassDeclaration -> {
+              env.logger.error(
+                "Only classes can be annotated with @ContributesBinding.", annotated
+              )
+              return@mapNotNull null
+            }
+            else -> annotated
+          }
+        }
+        .onEach {
+          it.checkClassIsPublic {
+            "${it.qualifiedName?.asString()} is binding a type, but the class is not public. " +
+              "Only public types are supported."
+          }
+          it.checkNotMoreThanOneQualifier(contributesBindingFqName)
+          it.checkSingleSuperType(contributesBindingFqName, resolver)
+          it.checkClassExtendsBoundType(contributesBindingFqName, resolver)
+        }
+        .forEach { clazz ->
+          val className = clazz.toClassName()
+
+          val scopes = clazz.getKSAnnotationsByType(ContributesBinding::class)
+            .toList()
+            .also { it.checkNoDuplicateScopeAndBoundType(clazz) }
+            .map { it.scope().toClassName() }
+            .distinct()
+            // Give it a stable sort.
+            .sortedBy { it.canonicalName }
+
+          val spec = generate(className, scopes)
+
+          spec.writeTo(
+            env.codeGenerator,
+            aggregating = false,
+            originatingKSFiles = listOf(clazz.containingFile!!)
```

Comment:
It's awesome that this is already built into the API (for both KSP and KotlinPoet) 🙌

Reply:
Yup one of the first things I implemented for KP's KSP support 😁